### PR TITLE
Optional recursive merge

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️

- https://github.com/yiisoft/app/pull/181
- https://github.com/yiisoft/app-api/pull/90
- https://github.com/yiisoft/demo/pull/334
- https://github.com/yiisoft/demo-api/pull/45

### CODE REVIEW

**Comment by vjik** [comment]
May be better add special method `Config::getWithRecursiveMerge()` ?

File: src/Config.php [hunk 1]
```diff
@@ -40,14 +41,24 @@ final class Config
...
         $this->rootPath = $rootPath;
         $this->relativeConfigsPath = trim($configsPath ?? Options::DEFAULT_CONFIGS_DIRECTORY, '/');
         $this->configsPath = $this->rootPath . '/' . $this->relativeConfigsPath;
+        $this->mergeRecursive = array_merge(
+            $mergeRecursive,
+            [
+                'params.php',
+                'events.php',
+                'events-web.php',
+                'events-console.php'
```
Comment (samdark) [inline src/Config.php:59]: That hard-codes some config names. We have `params` already but the rest wasn't forced by yiisoft/config previously. What if I want a config named `events` in my app but don't want to merge it recursively?

File: src/Config.php [hunk 2]
```diff
@@ -40,14 +41,24 @@ final class Config
...
-    public function __construct(string $rootPath, string $configsPath = null, string $environment = null)
+    public function __construct(string $rootPath, string $configsPath = null, string $environment = null, array $mergeRecursive = [])
     {
         $this->rootPath = $rootPath;
         $this->relativeConfigsPath = trim($configsPath ?? Options::DEFAULT_CONFIGS_DIRECTORY, '/');
         $this->configsPath = $this->rootPath . '/' . $this->relativeConfigsPath;
+        $this->mergeRecursive = array_merge(
+            $mergeRecursive,
+            [
+                'params.php',
```
Comment (samdark) [inline src/Config.php:56]: Is it necessary to specify `.php`? What if my `params` is mapped to `paramz.php` in `composer.json`?

File: src/Config.php [hunk 3]
```diff
@@ -237,7 +248,12 @@ private function merge(array $context, string $path = '', array ...$args): array
                         /** @var mixed */
                         $result[$k] = $v;
                     }
-                } elseif (is_array($v) && isset($result[$k]) && is_array($result[$k])) {
+                } elseif (
+                    in_array($context[0], $this->mergeRecursive, true) &&
```
Comment (samdark) [inline src/Config.php:252]: `in_array` check is O(n). If we'll reverse `$mergeRecursive` in constructor, i.e. will store it as:

```php
[
    'params.php' => 1,
    'events.php' => 1,
]
```

then we'll be able to use `array_key_exists()` or `isset()` and that would be O(1).

**Review by samdark — CHANGES_REQUESTED** [review]

**Comment by samdark** [comment]
What's the main reason to disable it by default? Performance?

**Comment by samdark** [comment]
OK. Got about the reason. The rest of questions remains.

File: src/Config.php (hunk 1 above)
Comment (yiiliveext) [reply src/Config.php:59]: Default parameters are your idea. In the original implementation, all configs for the recursive merge should be defined in the Config container (ApplicationRunner in the apps templates).

File: src/Config.php (hunk 2 above)
Comment (yiiliveext) [reply src/Config.php:56]: It is not necessary.

File: src/Config.php (hunk 3 above)
Comment (yiiliveext) [reply src/Config.php:252]: It is not a large array, you don't boost it much by this way. But you can do it if you think it's critical.

File: src/Config.php (hunk 1 above)
Comment (samdark) [reply src/Config.php:59]: He-he. I've changed opinion :)

File: src/Config.php (hunk 2 above)
Comment (samdark) [reply src/Config.php:56]: Let's remove it then.

**Pull request merged**